---
title: "DESeq2 Tutorial"
author: "Charlie George & Modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources

-   **DESeq2 Vignette** - <http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>
-   **Bioconductor RNAseq Workflow Book** - <http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html>

# install packages

-   If i didn't already have these installed in my environment - I would delete the \# on lines 19-23 and 26-39 and run the code

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

-   what do these packages do? Add comments to remind you what they do

```{r}
#A function to draw clustered heatmaps where one has better control over some graphical parameters such as cell size, etc.
library("pheatmap")   
#Creates nice looking color palettes especially for thematic maps
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)    
#Implements Empirical Bayes shrinkage and false discovery rate methods based on unimodal prior distributions.  
library(ashr)      
#apeglm provides Bayesian shrinkage estimators for effect sizes in GLM models, using approximation of the posterior for individual coefficients.
library(apeglm)  
#This is an organism specific package. The purpose is to provide detailed information about the species abbreviated in the second part of the package name org.Mm.eg.db. This package is updated biannually.
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

-   We made this (or a very similar version) using featurecounts in week1 on the cluster
-   raw data -\> fastqc -\> hisat2 -\> featurecounts

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv(
  "/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  rename_with(~str_remove(., '.bam')) %>% 
                  dplyr::select(!c("Chr","Start","End","Strand", "Length")) %>%
                  as.matrix()
counts_table


# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 

dim(counts_table)

```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv")
sample_info

# What class is the cell_type currently? 

sample_info$cell_type


# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 

sample_info$cell_type <- as.factor(sample_info$cell_type)


# what other columns might we want to include in our design that we want to change to factors? 

#genotype and replicate

sample_info$genotype <- as.factor(sample_info$genotype)
sample_info$replicate <- as.factor(sample_info$replicate)



```

# Check the sample names are the same and in the same order in the sample_info and counts_table

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 
identical(colnames(counts_table), sample_info$sample_id)

colnames(counts_table) == sample_info$sample_id
# make sample_info row names from sample_id 

#done by   rename_with(~str_remove(., '.bam')) %>% 
# remove .bam from colnames 


```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(countData = counts_table, 
                              colData = sample_info, 
                              design= ~genotype+cell_type)

ncol(counts_table)
nrow(sample_info)
  
```

## Investigate the dds object

```{r}
# what does it look like? - e.g. use str(dds)

str(dds)
# how to access counts

assays(dds)$counts

# access colData i.e the sample info table

colData(dds)
# access rowData

rowData(dds)

# access the design

design(dds)

# does it have any rowRanges()?

rowRanges(dds)

# what are its rownames and colnames?

row.names(dds)
colnames(dds)

# what attributes does it have?


```

## How to subset a DESeqDataSet Object

```{r}
# practice subsetting dds object
# select first 3 columns

dds[ , 1:3]


```

# Filter out unexpressed genes

-   It's good practice to filter out genes with 0 counts across all samples at the start
-   This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis
-   Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.

```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`
 dds_filtered <- dds %>% subset( rowSums(counts(dds))>0)
 
 dds_filtered <- dds[rowSums(counts(dds))>0, ]
 
# check dims of dds_filtered  
dim(dds_filtered)
```

# Perform exploratory analysis on dds_filtered object

1.  transform dds_filtered object using `rlog()` functions
    -   name your new variables `rlog_transformed_dds`
    -   we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    -   check the class of the rlog_transformed_dds objects
    -   you can look at the tranformed counts table using `assay()`

NOTE we only use this transformed data for QC not for downstream analysis or differential expression itself!!

```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind = TRUE)


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 

rlog_transformed_dds
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds)


```

# Optional: If you have time do the tranformation again but using the vst() function

-   As a guide
    -   rlog - more robust than VST when size factors vary widely (10x)
    -   VST - less sensitive to high count outliers than rlog, quicker - recommended most often

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- vst(dds_filtered, blind = TRUE)

vst_transformed_data

 assay(rlog_transformed_dds)

```

# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()

```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 

cor_matrix <- cor(assay(rlog_transformed_dds), method = "pearson")


# use pheatmap to plot the correlation matrix 

pheatmap(cor_matrix)

```

# try ploting a heatmap of sample distances

```{r}
# get sample distances -> Annotated what this code is doing?
#transpose gen-sample matrix to see distances of the samples not the genes - 
#t(assay(rlog_transformed_dds))

sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
        # cluster_rows = FALSE,
        # cluster_cols  = FALSE,
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data?

Answer: 2 sets of duplicates - ERR1755082 and ERR1755088 + ERR1755083 and ERR1755089

## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
colnames(rlog_transformed_dds)

rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6]

colnames(rlog_transformed_dds_subset)

 rlog_transformed_dds[,1:6]
# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above

sampleDists_subset <- dist(t(assay(rlog_transformed_dds_subset))) 
sample_distances_matrix_subset <- as.matrix(sampleDists) 

pheatmap(sample_distances_matrix_subset,
         clustering_distance_rows=sampleDists_subset,
         clustering_distance_cols=sampleDists_subset,
        # cluster_rows = FALSE,
        # cluster_cols  = FALSE,
         color = pretty_blue_colors)
```

What do you notice now about the data and the clustering?

Differenct cell types - 87, 85, 86 - cd8

# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r}
# plot pca of rlog_transformed_dds using plotPCA()

colData(rlog_transformed_dds_subset)

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type") + coord_fixed(ratio = 10)

# what does the intgroup parameter do? have a go at changing it...

plotPCA(rlog_transformed_dds_subset, intgroup = "replicate") + coord_fixed(ratio = 10)


# how many variable genes is it using for pca? 

#500 genes by default of ntop in function


# plot the pca for all genes 

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = nrow(rlog_transformed_dds_subset)) + coord_fixed(ratio = 5)


```

What does it tell us about our sample?

Answer:

------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis e.g. - update design with variables in sample info table we want to regress out - remove samples (poor quality or those with name switches etc)

# Revisiting our initial data

What do we need to do to our data???

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
dds_filtered_updated <- dds_filtered[,1:6]
# do we want to update the design? Is this nessecary in this case?

design(dds_filtered_updated) <- ~cell_type

```

# Now we can start our actual DE!!!! yay

Run deseq2 modelling using one function (DESeq())

```{r}

# look at dds slots then run deseq2 function 
# what has happened to your dds slots?

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

mcols(results_obj)$description

# sumarise results 
summary(results_obj)

```

Have a look at the outputs

# Now we will run the analysis step by step to see whats going on

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()

dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)
# Access the size factors from the dds_filtered_updated object

sizeFactors(dds_filtered_updated)

# Do you see any systematic differences in size factors between groups?


```

# Optional: Plot size factors as barchart using ggplot

```{r}
size_factors <- sizeFactors(dds_filtered_updated)
dds_filtered_updated_df <- as.data.frame(size_factors)

 dds_filtered_updated_df <- rownames_to_column(dds_filtered_updated_df, var = "sample")  
  dds_filtered_updated_df <- cbind(dds_filtered_updated_df, sample_info[1:6, ])
  
ggplot( dds_filtered_updated_df, aes(x = sample, y = size_factors, fill = cell_type) ) + 
         geom_col()
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts
 normalized_counts <- counts(dds_filtered, normalized = TRUE)

# check the class of the output
 
class(normalized_counts)

head(normalized_counts)
```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 

new <- estimateDispersions(dds_filtered_updated)

# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts(new)
```

## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest

wald_test <- nbinomWaldTest(dds_filtered_updated)
head(wald_test)
summary(results(wald_test))
summary(wald_test)
wald_test

```

# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# generate dds object 

featurecounts_table <- read_tsv(
  "/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  rename_with(~str_remove(., '.bam')) %>% 
                  dplyr::select(!c("Chr","Start","End","Strand", "Length")) %>%
                  as.matrix()

sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv")

sample_info$cell_type <- as.factor(sample_info$cell_type)

dds <- DESeqDataSetFromMatrix(countData = counts_table, 
                              colData = sample_info, 
                              design= ~genotype+cell_type)

ncol(counts_table)
nrow(sample_info)


# filter rows with 0 counts 

 dds_filtered <- dds %>% subset( rowSums(counts(dds))>0)

# remove unwanted samples 
 
 dds_filtered <- dds_filtered[,1:6]
 design(dds_filtered) <- ~cell_type


# make DESeqDataSet (dds) object 
 # run DESeq() 

 
 dds_filtered <- DESeq(dds_filtered)
 

```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 

results_obj2 <- results(dds_filtered)
mcols(results_obj2)$description

head(results_obj2)
summary(results_obj2)
summary(dds_filtered)
summary(results(dds_filtered))

```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM -- see vingette("DESeq2")

```{r}

```

# Plot normalised counts for ENSMUSG00000041779

checkout the 'plotCounts()\` function

```{r}
# Plot normalised counts for gene ENSMUSG00000041779

plotCounts(dds_filtered, "ENSMUSG00000041779", intgroup = "cell_type")
```

# Access the results table for the comparison between CD8 and CD4 T cells . You can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control

checkout `resultsNames()` & `results()`

```{r}

resultsNames(dds_filtered)

results <- results(dds_filtered, contrast = c("cell_type", "CD4", "CD8"))

```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot -- do they look as expected? for CD8 vs CD4

```{r}

results_df <- as.data.frame(results)

results_df <- rownames_to_column(results_df, var = "sample")

ggplot(results_df) +
  geom_histogram(aes(x = pvalue, fill = "blue")) +
  geom_histogram(aes(x=padj, fill = "red",  alpha = 0.49))

```

# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}

plotMA(dds_filtered,  alpha = 0.05)

```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -\> see the vignette -- which shall we choose?

```{r}

# call your results table  called shrunkenLFC_dds_results_obj

# call your results table  called shrunken_dds_results_df
shrunken_dds_results_df <- lfcShrink(dds_filtered, coef = "cell_type_CD8_vs_CD4", type="apeglm")

#using the DEseq MA plot function
plotMA(shrunken_dds_results_df, alpha=0.05, colNonSig = "gray60",
  colSig = "blue",
  colLine = "black")

#plotting the same graph in ggplot
shrunken_dds_results_df %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

 


#shrunkenLFC_dds_results_obj <-lfcShrink(dds_filtered, coef = "cell_type_CD8_vs_CD4", type = "ashr")
#plotMA(shrunkenLFC_dds_results_obj,  alpha = 0.05)


```

# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df

shrunkenLFC_dds_results_df <- as.data.frame(shrunken_dds_results_df)

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 

MGI_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_df), "MGI", "ENSEMBL")
gene_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_df), "SYMBOL", "ENSEMBL")
entrez_ids <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_df), "ENTREZID", "ENSEMBL")

table(names(entrez_ids)==row.names(shrunken_dds_results_df))     #to check that the order of the vectors are the same as the df
#options(scipen=0)

#add mapId vectors to the original results df
results_shrunken_LFC_filtered_df <- shrunken_dds_results_df %>%
    as.data.frame() %>% 
    mutate(mgi_symbols=MGI_symbols) %>%
    mutate(gene_symbols=gene_symbols) %>%
    mutate(ENTREZ_IDs=entrez_ids)


# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?
which(duplicated(names(MGI_symbols)))

```


## Exportthe results to CSV

```{r}
#results_shrunken_LFC_apeglm_filtered_df <- results_shrunken_LFC_apeglm_filtered_df %>% 
#						rownames_to_column("ensembl_gene_ids")
						
write_csv(results_shrunken_LFC_filtered_df, "CD8_vs_CD4_DE_results_lfcshrinkage.csv")
```

```

# Export all the results to CSV

```{r}
write_csv(shrunkenLFC_dds_results_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```

# Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```

Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account.

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....


#paste code
```{r}
# filter rows with 0 counts 
dds_actual_analysis <- dds[rowSums(counts(dds)) > 0, ]


# remove unwanted samples 
dds_actual_analysis <- dds_actual_analysis[,1:6]

# make DESeqDataSet (dds) object 
design(dds_actual_analysis) <- ~ cell_type

# run DESeq() 
dds_actual_analysis <- DESeq(dds_actual_analysis)
```

# 7. Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 
results_obj <- results(dds_actual_analysis, alpha=0.05)
summary(results_obj)
```

## Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}

head(coef(dds_actual_analysis))
coef_dds <- coef(dds_actual_analysis)

# Intercept is absolute gene expression in baseline group in rea
# Coefficients are the logFC in the other conditions
# Coefficients are the log2 fold changes for each group

##### EXTRA #############################

# Convert coefficients to "real" space rather than log values
coef_real <- 2^(coef(dds_actual_analysis))
coef_real["ENSMUSG00000041779", ]
# Intercept     376.604199
# cell_type_CD8_vs_CD4    1.251456

# Intercept - log2 normalised expression level of gene for baseline condition, which is Egr2_3_DKO_CD4 (uses size factors and dispersion estimates calculated earlier)
```

# Plot normalised counts for gene ENSMUSG00000041779

```{r}

gene_counts_filt <- counts(dds_actual_analysis, normalized = TRUE)
gene_counts_filt <- gene_counts_filt["ENSMUSG00000041779", ]
gene_counts_filt <- data.frame(counts = gene_counts_filt,
                               condition = dds_actual_analysis$cell_type)
gene_counts_filt
ggplot(gene_counts_filt, aes(x = condition, y = counts)) +
    geom_boxplot() +
    geom_point()
# Coefficients above show that gene is expressed at 1.25 times the background level in Egr2/3 DKO CD8s (control is Egr2/3 DKO CD4)
```
```{r}
# Another way to do the same plot using plotCounts functon to returnData 
#plot_data <- plotCounts(dds = dds_actual_analysis, gene = "ENSMUSG00000041779", intgroup = c("cell_type","replicate"), returnData = T)
#ggplot(plot_data, aes(x = cell_type, y = count, group=cell_type)) +
 #   geom_boxplot() +
  #  geom_point(aes(color=replicate), size=2)
```

## Access the results table for the comparison between CD8 and CD4 T cells

```{r}
resultsNames(dds_actual_analysis)
results_CD4vsCD8 <- results(dds_actual_analysis,contrast = c('cell_type','CD4','CD8'))
summary(results_CD4vsCD8)


results_CD8vsCD4 <- results(dds_actual_analysis,contrast = c('cell_type','CD8','CD4'))
summary(results_CD8vsCD4)


results_rep1vsrep3 <- results(dds_actual_analysis,contrast = c('replicate','1','3'))
summary(results_rep1vsrep3)
```

Table should contain log2 fold changes and p-values for all genes

## Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected?

```{r}
results_obj %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_histogram(aes(x=pvalue), fill='lightblue') + 
    geom_histogram(aes(x=padj),fill='lightgreen')

```


## Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(results_obj,alpha=0.05, colNonSig = "gray60")

results_obj %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 
results_obj


```

## Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunken_dds_results_df
shrunken_dds_results_apeglm <- lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="apeglm")

#using the DEseq MA plot function
plotMA(shrunken_dds_results_apeglm, alpha=0.05, colNonSig = "gray60",
  colSig = "blue",
  colLine = "black",)

#plotting the same graph in ggplot
shrunken_dds_results_apeglm %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

```


```{r}
# test otehr shrinkage methods
#plotting the same graph in ggplot
lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="normal") %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

# test otehr shrinkage methods
# plotting the same graph in ggplot
lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="ashr") %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

```

## Generate a results table (one shrinkage method) containing mgi symbols

```{r}
# Use the org.Mm.eg.db package (hint:you did this yesterday)
# Remove all genes with a padj of NA
# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?

keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# # mapIds gives you a 1:1 mapping by default (can only extract one column at a time), whereas the select function doesn't (but you can extract multiple columns at a time)

MGI_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "MGI", "ENSEMBL")
gene_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "SYMBOL", "ENSEMBL")
entrez_ids <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "ENTREZID", "ENSEMBL")

table(names(entrez_ids)==row.names(shrunken_dds_results_apeglm))     #to check that the order of the vectors are the same as the df
#options(scipen=0)

#add mapId vectors to the original results df
results_shrunken_LFC_apeglm_filtered_df <- shrunken_dds_results_apeglm %>%
    as.data.frame() %>% 
    mutate(mgi_symbols=MGI_symbols) %>%
    mutate(gene_symbols=gene_symbols) %>%
    mutate(ENTREZ_IDs=entrez_ids)


# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?
which(duplicated(names(MGI_symbols)))

```


## Exportthe results to CSV

```{r}
results_shrunken_LFC_apeglm_filtered_df <- results_shrunken_LFC_apeglm_filtered_df %>% 
	rownames_to_column("ensembl_gene_ids")
						

write_csv(results_shrunken_LFC_apeglm_filtered_df, "CD8_vs_CD4_DE_results_lfcshrinkage.csv")

```

## Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....


